src/check_python_versions/sources/poetry_pyproject.py

```diff
@@ -0,0 +1,197 @@
+    classifiers = _get_pyproject_toml_classifiers(filename)
+
+    if classifiers is None:
+        # Note: do not return None because pyproject.toml is not an optional source!
+        # We want errors to show up if pyproject.toml fails to declare Python
+        # versions in classifiers.
+        return []
+
+    return get_versions_from_classifiers(classifiers)
+
+
+def get_python_requires(
+    pyproject_toml: FileOrFilename = PYPROJECT_TOML,
+) -> Optional[SortedVersionList]:
+    """Extract supported Python versions from python_requires in pyproject.toml."""
+    python_requires = _get_pyproject_toml_python_requires(pyproject_toml)
+    if python_requires is None:
+        return None
+    if not isinstance(python_requires, str):
+        warn('The value passed to python is not a string')
+        return None
+    return parse_python_requires(python_requires)
+
+
+def update_supported_python_versions(
+    filename: FileOrFilename,
+    new_versions: SortedVersionList,
+) -> Optional[FileLines]:
+    """Update classifiers in a pyproject.toml.
+
+    Does not touch the file but returns a list of lines with new file contents.
+    """
+    classifiers = _get_pyproject_toml_classifiers(filename)
+    if classifiers is None:
+        return None
+    if not isinstance(classifiers, (list, tuple)):
+        warn('The value passed to setup(classifiers=...) is not a list')
+        return None
+    new_classifiers = update_classifiers(classifiers, new_versions)
+    return _update_pyproject_toml_classifiers(filename, new_classifiers)
+
+
+def update_python_requires(
+    filename: FileOrFilename,
+    new_versions: SortedVersionList,
+) -> Optional[FileLines]:
+    """Update python dependency in a pyproject.toml, if it's defined there.
+
+    Does not touch the file but returns a list of lines with new file contents.
+    """
+    python_requires = _get_pyproject_toml_python_requires(filename)
+    if python_requires is None:
+        return None
+    comma = ', '
+    if ',' in python_requires and ', ' not in python_requires:
+        comma = ','
+    space = ''
+    if '> ' in python_requires or '= ' in python_requires:
+        space = ' '
+    new_python_requires = compute_python_requires(
+        new_versions, comma=comma, space=space)
+    if is_file_object(filename):
+        # Make sure we can read it twice please.
+        # XXX: I don't like this.
+        cast(TextIO, filename).seek(0)
+    return _update_pyproject_toml_python_requires(filename, new_python_requires)
+
+
+def _update_pyproject_toml_classifiers(
+        filename: FileOrFilename,
+        new_value: Union[str, List[str]]
+) -> Optional[FileLines]:
+    with open_file(filename) as f:
+        table = pytomlpp.load(f)
+
+    if 'tool' not in table:
+        return []
+    if 'poetry' not in table['tool']:
+        return []
+
+    table['tool']['poetry']['classifiers'] = new_value
+
+    return pytomlpp.dumps(table).split('\n')
```

Comment:
Ah!  Does this preserve formatting/comments?  This would be a sufficiently good reason not to use tomli.

Reply:
I've done a test with my pyproject, and the dumped content is different than original one.
I've searched pytomlpp and seems there's no way to keep formatting/comments (and also the original toml style).

that dumps is used to create a big string with the content of entire file.

Reply:
Switched to tomlkit that keeps the original style and comments.